### PR TITLE
Styling guide worked on extensively -- first draft

### PR DESCRIPTION
**Pull request by mercmobily**
Following up from #1533 , I decided that asking questions and asking you guys to "rewrite a section" was less than ideal .So,I sat down over the week end (well, actually, into Monday morning) and worked on an extensive rewrite.

After wasting some of your time, hopefully you will find here something valuable to use or take apart.

In general:

 * I made the examples more uniform and more comprehensive, always using `<my-element>` and `<my-toolbar>`

* I restructured it a little, so that there is always a "normal" explanation and then a section on the shim's limitations.

* Whenever possible, I tried my best to explain things (like the dynamism and inheritance limitations) using code that matched what the reader had read previously in the chapter

* I explained inheritance limitations providing -- again -- an example similar to the previous ones, but adding a <span> to the mix AND (not sure here if it made sense) trying to explain, "give or take", what happens behind the scenes and emphasising that Polymer elements are treated differently.

As usual: no hard feelings if you scrap it all. It was a good exercise for me anyway.

### CODE REVIEW

**Comment by arthurevans** [comment]
Heh. Ironically, I started to do some rewrites on the plane home. Let me take a look your changes and figure out what we should do.

File: 1.0/docs/devguide/styling.md [hunk 1]
```diff
@@ -8,11 +8,27 @@ subtitle: Developer guide
 
 {% include toc.html %}
 
-
 Polymer uses [Shadow DOM styling
 rules](http://www.html5rocks.com/en/tutorials/webcomponents/shadowdom-201/) for
-providing scoped styling of the element's local DOM.  Scoped styles should be
-provided via `<style>` tags placed inside the element's local DOM `<template>`.
+providing scoped styling of the element's local DOM. This allows elements to
+create style rules that will only apply to the element's light DOM or local
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: 2 points here:

-   "light DOM" is not correct. This should probably be "the element's local DOM and distributed children." 

    Explanation: there are only two, specific places where scoped rules can touch stuff outside the element's local DOM: you can style the host element itself, with `:host`, and you can style distributed children, with `::content`. The distinction between light DOM children and distributed children is subtle, but important. If you don't have a `<content>` tag, you can't style your light DOM children, for instance. If you use `<content select=`, you can't style anything that doesn't match one of your selectors, because it doesn't get distributed.

-   The CSS custom properties shouldn't be mentioned in the same sentence. They're not related to the shadow DOM spec, they're a completely different specification that happens to be useful for styling shadow DOM. (We also talk about "the shady DOM shim" and "the custom properties shim" as separate entities.)

File: 1.0/docs/devguide/styling.md [hunk 2]
```diff
@@ -8,11 +8,27 @@ subtitle: Developer guide
...
+DOM, as well as advanced theming with CSS custom properties and CSS mixins.
+
+Scoped styles should be provided via `<style>` tags placed inside the element's
+local DOM `<template>`.
+
+To place styles outside of the element, or share styles between elements, you
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: This line is not quite right. Style modules are for sharing styles, yes. The `custom-style` element is what you use for defining local DOM aware style rules outside of an element.

File: 1.0/docs/devguide/styling.md [hunk 3]
```diff
@@ -8,11 +8,27 @@ subtitle: Developer guide
...
+can create a [style module](#style-modules).
+
+**NOTE: I think a nice introduction, without code, works a lot better. I also added a brief explanation.**
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: Yeah, I agree. 

(Longer-term plan, BTW, is to separate conceptual docs ("How shadow DOM scoping works") from task-oriented docs ("Style a local DOM element"). But working on that ATM for data binding; styling is probably a ways off.

File: 1.0/docs/devguide/styling.md [hunk 4]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
             display: block;
             border: 1px solid red;
           }
-          #child-element {
+          /* In shady DOM this will become `.child-element.my-element {` */
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: We haven't really talked about how shady DOM shims things yet, so it's a little confusing to say, "In shady DOM this becomes ..." This is an implementation detail of how the shady DOM styling works, so if we discuss it, it should be in the context of the shim implementation/shim limitations.

File: 1.0/docs/devguide/styling.md [hunk 5]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
...
+        <div class="special">This should be special!</div>
+    </my-element>
 
-        x-foo div
+Only `This should be special!` will have an orange background.
 
-    (Where `x-foo` is the name of the custom element.)
+#### Shim limitations
 
-*   To limit styles to elements inside the ::content tag, add a wrapper element around the
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: This deleted paragraph concisely describes why we're using the workaround starting at L133.

File: 1.0/docs/devguide/styling.md [hunk 6]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
...
-    `<content>` element. This is especially important when using a child combinator (`>`) to
-    select top-level children.
+Shady DOM is much more limited than shadow DOM.
 
-        <dom-module id="my-element">
+Under shady DOM,  the `<content>` tag doesn't appear in the DOM tree, and styles
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: It is really confusing to insert this paragraph here just before describing a different limitation, which _also_ involves the child combinator. (This paragraph talks about `:host > ::content foo`, as opposed to `:host ::content > foo`, which are different selectors and see different transforms.

Honestly I think this is a much rarer case and I would not put it first.

File: 1.0/docs/devguide/styling.md [hunk 7]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
...
+are rewritten to remove the `::content` pseudo-element **as well as and any
+combinator immediately to the left of `::content`.** (E.g. `> :: content` where `>`
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: E.g. => for example (we avoid Latin).

File: 1.0/docs/devguide/styling.md [hunk 8]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
...
+is rewritten as `my-element > .special`. This means that the rule will not apply
+just to the light DOM (as it should), but to the local DOM too (which it
+shouldn't).
 
-            <style>
-              .content-wrapper ::content > .special {
-                background: orange;
-              }
-            </style>
+**NOTE: The original documentation is very confusing here, as it says that you must have a selector before `::content` (for example `:host`), but in my experience that doesn't actually do anything since BOTH `:host ::content div` and `::content div` becomes `element div` (which is obviously not enough). I wonder if adding `element` was a recent-ish addition, and earlier versions of polymer had `::content div` turn into a dangerous, leaky `div`!**
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: Yeah, I think  the doc is confusing here. Several things got conflated. 

At one point, you just needed a selector there.

But if you're using a child combinator, or if you have anything else in the local DOM that could be affected by these styles, you should use the wrapper approach. 

You highlighted one of the possible issues, here's the other one:

Rewriting `::content > .special` as `my-element > special` is potentially wrong because `>` selects _direct children`. So consider this element template:

    <template>
      <div id="main">
        <content></content>
      </div>
    </template>

Used like this:

    <my-element>
      <div class="special">I'm special!</div>
      <div>I'm not.</div>
    </my-element>
      
If I use the rule `::content > .special`, and in shady DOM that gets transformed to `my-element > .special` ... And it no longer selects anything, because the _rendered_ tree looks like this:

    <my-element>
       <div id="main">
         <div class="special">I'm special!</div>
         <div>I'm not.</div>
       </div>
    </my-element>

So, short version: If you have a very simple local DOM template like this:

    <template>
      <content></content>
    </template>

You can get away with just `::content selector`.

If your element has other local DOM contents that could match `selector`, or if you want to select only the top-level distributed children (using the child combinator (`>`)), then you need to use a wrapper.

Comment (mercmobily) [inline 1.0/docs/devguide/styling.md]: I think this should be (nicely) explained in the documentation. I mean, it looks like too much detail, but.. isn't that where the devil is? I will be happy to include this in the explanation, if you're OK with it.

File: 1.0/docs/devguide/styling.md (hunk 5 above)
Comment (mercmobily) [inline 1.0/docs/devguide/styling.md]: I know... my point being "too concisely?". Anyway, let me know if you'd like to keep that line :D

File: 1.0/docs/devguide/styling.md (hunk 4 above)
Comment (mercmobily) [inline 1.0/docs/devguide/styling.md]: Very true.

File: 1.0/docs/devguide/styling.md (hunk 1 above)
Comment (mercmobily) [inline 1.0/docs/devguide/styling.md]: Yes, yes and yes.

File: 1.0/docs/devguide/styling.md (hunk 7 above)
Comment (mercmobily) [inline 1.0/docs/devguide/styling.md]: OK

**Comment by arthurevans** [comment]
Thanks for all this work, Merc.

Not going to finish reviewing today. Let me finish reviewing and think this whole thing over before you put too much more time into this. For the custom properties shim in particular, I think we can make things a lot clearer with some structural changes.

**Comment by mercmobily** [comment]
OK no worries. I was hoping to create something "nearly finished" -- and obviously failed again! I hope I didn't just create extra work in the process of trying to help...

File: 1.0/docs/devguide/styling.md [hunk 9]
```diff
@@ -342,112 +421,195 @@ Example usage of `my-toolbar`:
...
-Below are current limitations of the shim. Improvements to performance and
-dynamism will continue to be explored.
+For the following example, adding/removing the `urgent` class on the
+`<my-toolbar>` element will have the desired effect, since the dynamism is
+related to *application* of a custom property.
 
-#### Dynamism limitations
+**NOTE: as a selector, I used `:host.urgent span.title` rather than
+**:host(.urgent) span.title mainly because the :host(.selector) syntax hasn't
+**been used at all in the guide before now. I hope it's OK and it's "formally"
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: `:host(.urgent)` is correct. `:host.urgent` won't work under shadow DOM (and probably shouldn't work under shady DOM).

File: 1.0/docs/devguide/styling.md [hunk 10]
```diff
@@ -342,112 +421,195 @@ Example usage of `my-toolbar`:
...
+You can directly modify a {{site.project_title}} element's custom property by
+setting key-value pairs in
+[`customStyle`](../../api/#Polymer.Base:property-customStyle) on the element.
+This is analogous to setting `style` on an element.
 
-You can  directly modify a {{site.project_title}} element's custom property by setting
-key-value pairs in [`customStyle`](../../api/#Polymer.Base:property-customStyle) on the
-element (analogous to setting `style`) and then calling `updateStyles`. Or you can pass
-a dictionary of property names and values as an argumen to `updateStyles`.
+Note that in shady DOM you then need to call `updateStyles()` in order for the
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: This is a limitation of the custom properties shim and is not related to shady DOM.

File: 1.0/docs/devguide/styling.md [hunk 11]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
...
+    .content-wrapper ::content > .special
 
-        .content-wrapper > .special
+Becomes:
 
+    .content-wrapper.my-element  > .special
 
+**NOTE: I have a bit of a problem here as the chapter hasn't yet introduced**
+**custom properties, and yet it warns "Custom properties can't style distributed**
+**children. I am commenting this warning out for now**
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: This is true, but it's also true that people don't tend to read docs in order. This is a large enough limitation (and a big enough source of confusion) that I wanted to put it up front here.

**Comment by mercmobily** [comment]
Shall I get started on this Arthur?

File: 1.0/docs/devguide/styling.md [hunk 12]
```diff
@@ -8,11 +8,27 @@ subtitle: Developer guide
...
+
+Scoped styles should be provided via `<style>` tags placed inside the element's
+local DOM `<template>`.
+
+To place styles outside of the element, or share styles between elements, you
+can create a [style module](#style-modules).
+
+**NOTE: I think a nice introduction, without code, works a lot better. I also added a brief explanation.**
+
+# Styling the local DOM
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: This should be an H2 (`##`).

File: 1.0/docs/devguide/styling.md [hunk 13]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
...
+And see that only `In local DOM` has a yellow background.
 
 **Note:**  Prior to Polymer 1.1, the recommendation was to place `<style>` tags
 inside the `<dom-module>` for an element (but _outside_ the `<template>`). This
  is still supported, but is no longer recommended.
 {: .alert .alert-info }
 
+**NOTE: I like introducing what shadow DOM styling SHOULD be like, and then explain what we should do with shady DOM, and why.**
+
+#### Shim limitations
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: Like these, but how about renaming them to "Shady DOM limitations" since we have both shady DOM and the custom properties shims to discuss in this chapter.

**Comment by arthurevans** [comment]
Hi Merc,

Sorry to be slow on this, but I was out of the country for meetings last week and got called up for jury duty this week, so I'm way, way behind.

I don't have the bandwidth to deal with the custom properties stuff right now. I think it needs more structural work. 

Could you break out the first half of the PR? I think we're closer on that, and it's a more manageable chunk to review.

File: 1.0/docs/devguide/styling.md [hunk 14]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
...
+
+There is no need to do any special workaround in this simple case, since shady
+DOM will automatically change the style as well as the element's class to make
+sure that styling in the local DOM doesn't leak out of `my-element` by adding a
+required class, `my-element`, to the created element (and to the selector).
+
+
+### Styling distributed children (`::content`)
+
+In shadow DOM, you can style an element's light DOM using the `::content` pseudo element. So, you can do this:
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: again, light DOM => distributed children throughout. pseudo-element is hyphenated.

File: 1.0/docs/devguide/styling.md [hunk 15]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
...
 inside the `<dom-module>` for an element (but _outside_ the `<template>`). This
  is still supported, but is no longer recommended.
 {: .alert .alert-info }
 
+**NOTE: I like introducing what shadow DOM styling SHOULD be like, and then explain what we should do with shady DOM, and why.**
+
+#### Shim limitations
+
+There is no need to do any special workaround in this simple case, since shady
+DOM will automatically change the style as well as the element's class to make
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: Run-on sentence. Maybe something like:

Shady DOM emulates shadow DOM style encapsulation by adding classes to elements inside the local DOM tree at runtime, and by rewriting, or _shimming_ the style rules. For example, shady DOM adds the classes `style-scope` and `my-element` to the local DOM tree for `<my-element>`. The selector `.child-element` is rewritten as `.my-element.child-element`.

Shady DOM only shims style rules that appear _inside_ an element's local DOM, or in a [`custom-style` element ](link forward). Styles defined at the document level inside a normal `<style>` tag are **not** shimmed and can leak into local DOM trees. To avoid this issue, use a `custom-style` element for document-level styling.

File: 1.0/docs/devguide/styling.md [hunk 16]
```diff
@@ -24,92 +40,148 @@ provided via `<style>` tags placed inside the element's local DOM `<template>`.
...
+          });
+      </script>
 
-This implies:
+    </dom-module>
 
-*   You must have a selector to the left of the `::content` pseudo-element.
 
-        :host ::content div
+In shadow DOM, the `.special` class will only ever affect elements in the light
```
Comment (arthurevans) [inline 1.0/docs/devguide/styling.md]: `.special` class => the `::content .special` selector

We use present tense (almost) everywhere, so "will only ever affect elements in the light DOM" => only matches distributed children. If you want to emphasize this more, could be "matches distributed children and _only_ distributed children."

**Comment by mercmobily** [comment]
I am back... I locked myself in a cave till I finished the server side of the story of my project.
Now... I am eager to get this finished. Arthur, what would you like me to do? Shall I try and apply _all_ of the request you made, and see where we are at?

(Hopefully, at that point the review on your side won't take as much energy...)

I am happy to pounce on it whenever you want me to.

**Comment by mercmobily** [comment]
(I will fix the conflicts too, the file has had a few commits since I made this PR)

**Comment by arthurevans** [comment]
Welcome back @mercmobily! I was afraid I'd scared you off entirely.

Unfortunately, I've recently entered my own cave over here. If you can do just the first half (overview portion) in the next couple of days I can review it. Otherwise, I'm not going to get a look at it until after IO and I can guarantee more merge conflicts at that time :).

**Comment by mercmobily** [comment]
Hi,

Damn. I was away those couple of days... I am too late right? :(
I am happy to go through it and resolve any conflicts really...

I think it's best if we wait till after IO.

Merc.
ᐧ

**Comment by arthurevans** [comment]
Hi @mercmobily ... the new site is finally on the master branch here, so we had to close this (more explanation below):

Sorry we had to close this PR. We replaced the master branch of the docs repo, with the all-new docs site. Because there's no shared history with the old master branch, the easiest way to get started with the new site is to clone a new repo and follow the install instructions in the [README](https://github.com/Polymer/docs/blob/master/README.md). Thanks for your contribution, and apologies for the inconvenience.

**Comment by mercmobily** [comment]
Hi,

No worries, thank you Arthur!

merc.
